src/edgeiter.jl

```diff
@@ -7,80 +7,47 @@ end
-end
-
-function EdgeIter(g::DiGraph)
-    di = 1
-    s = 1
-    ne(g) == 0 && return EdgeIter(ne(g), g.fadjlist, EdgeIterState(0,0,true), false)
-    while di > length(g.fadjlist[s])
-        s += 1
-        di = 1
-        s > length(g.fadjlist) && return EdgeIter(ne(g), g.fadjlist, EdgeIterState(1,1, true), true)
+function _next(eit::EdgeIter, state::EdgeIterState = EdgeIterState(1,1,false))
+    while state.s <= length(eit.adj)
+        arr = eit.adj[state.s]
+        while state.di <= length(arr)
+            if eit.directed || state.s <= arr[state.di]
+                return state
+            end
+            state.di += 1
+        end
+        state.s += 1
+        state.di = 1
     end
-    return EdgeIter(ne(g), g.fadjlist, EdgeIterState(s, di, false), true)
+    state.fin = true
+    return state
 end
 
-
-start(eit::EdgeIter) = eit.start
+start(eit::EdgeIter) = _next(eit)
 done(eit::EdgeIter, state::EdgeIterState) = state.fin
 
-_isfin(eit::EdgeIter, state::EdgeIterState) =
-    state.s > length(eit.adj) || (
-        state.s == length(eit.adj) &&
-        state.di > length(eit.adj[state.s])
-    )
-
-function next(eit::EdgeIter, state::EdgeIterState)
-    # calculate the edge we're currently looking at.
-    # this is guaranteed to be valid.
-    d = eit.adj[state.s][state.di]
-    edge = Edge(state.s, d)
-    found = false       # have we found a valid next state?
-
-    # now, let's get the next valid state, or set fin if there are no more.
+function next(eit::EdgeIter, state)
+    edge = Edge(state.s, eit.adj[state.s][state.di])
+    state.di += 1
+    return(edge, _next(eit, state))
+end
 
-    while !found
-        state.di += 1                           # increase di.
-        while (state.s < length(eit.adj) &&     # if we're at the end of a vector and
-            state.di > length(eit.adj[state.s]))  # not at the end of the list
-            # println("end of vector $(state.s)")
-            state.s += 1                        # go to the next vector, and
-            state.di = 1                        # index to the first element.
-        end
-        if _isfin(eit, state)                   # oops, we've hit the end
-            state.fin = true
-            return(edge, state)             # return a finished nextstate
-        end
-        if !eit.directed                    # for undirected graphs
-            if state.s <= eit.adj[state.s][state.di]     # skip edges where s > d
-                found = true
-            # else
-            #     println("skipping because $(state.s) > $(eit.adj[state.s][state.di])")
-            end
-        else
-            found = true
-        end
-    end
-    state.fin = _isfin(eit, state)                   # oops, we've hit the end
-    return (edge, state)
+function collect(eit::EdgeIter)
+   es = sizehint!(Edge[], eit.m)
+   for u in 1 : length(eit.adj)
+      for v in eit.adj[u]
+         push!(es, Edge(u,v))
+      end
+   end
+   return es
```

Comment:
`Collect` is already defined since the object is iterable. But its likely to be slower, and will result in a few more allocations. I'll make the change for undirected graphs.